gradle/release.gradle

```diff
@@ -106,121 +105,147 @@ task("releaseNeeded") {
-        }
-        run pushCommandLine
+task bumpVersion {
+    description = "Increments version in 'version.properties' and commits to '$System.env.TRAVIS_BRANCH'"
+    doLast {
+        String nextVersion = project.versionFile.incrementVersion()
+        String message = "Increment version '$currentVersion' -> '$nextVersion' $buildInfo"
+        run "git", "commit", "-m", "$message" as String, "version.properties"
     }
 }
 
-/*
-TODO:
- - declare release steps and then Gradle plugin adds tasks:
-  release, rollbackRelease, releaseStep1, rollbackStep2, releaseSteps1..7
- - improve exception reporting
- - start rolling back from last successful step (e.g. don't rollback the step that failed)
- */
-
-private void commitReleaseNotes(String buildInfo) {
-    def notesFile = project.file("doc/release-notes/official.md")
-    run "git", "add", "$notesFile" as String
-    run "git", "commit", "-m", "Update release notes $buildInfo" as String, "$notesFile" as String
+task unbumpVersion(type: Exec) {
+    description = "Unbumps version in 'version.properties' by resetting head by one commit (--hard)."
+    commandLine "git", "reset", "--hard", "HEAD~1"
 }
 
-private void createTag(String buildInfo, String tag) {
-    String tagMessage = "Create tag ${tag} ${buildInfo}"
-    run "git", "tag", "-a", tag, "-m", "$tagMessage" as String
+task pushChanges(type: Exec) {
+    description = "Pushes changes to '${System.env.TRAVIS_BRANCH}' branch."
+    //!!!Below command _MUST_ be quiet otherwise it exposes GitHub write token!!!
+    def mustBeQuiet = "-q"
+    def pushCommandLine = ["git", "push", "https://szczepiq:${System.env.GH_TOKEN}@github.com/mockito/mockito.git", System.env.TRAVIS_BRANCH, "v$currentVersion", mustBeQuiet]
+    if (releaseTest) {
+        pushCommandLine << '--dry-run'
+    }
+    commandLine pushCommandLine
 }
 
-private void commitIncrementedVersion(String currentVersion, String buildInfo, VersionFile versionFile) {
-    String nextVersion = versionFile.incrementVersion()
-    String message = "Increment version '$currentVersion' -> '$nextVersion' $buildInfo"
-    run "git", "commit", "-m", "$message" as String, "version.properties"
+releaseWorkflow {
+    step assertCleanWorkingCopy
+    step releaseNeeded
+    onlyIf { releaseNeeded.needed }
+    step bintrayUpload
+    step pullCommits
+    step configureGenericGitUser, [cleanup: restoreGitUser]
+    /*
+        Now we'll start operating on branch that we want to release.
+        This introduces a problem - someone might have pushed changes *after* release process has started
+        What can happen:
+         - bintrayUpload will fail saying that the version is already released
+         - git push fails saying that the pull is needed
+        Resolutions:
+         - see 'TROUBLESHOOTING' section at the top of this file
+    */
+    step checkOutCommittableBranch
+    step updateReleaseNotes
+    step commitReleaseNotes, [rollback: resetReleaseNotes]
```

Comment:
Yes, we could do that. We could make the upload to be the very last step. I think this is a very good idea. I implement it outside of this PR. This PR only refactors existing behavior. I'll add a TODO.